internal/credential/vault/jobs.go

```diff
@@ -0,0 +1,200 @@
+	const op = "vault.NewNewTokenRenewal"
+	if repoFn == nil {
+		return nil, errors.New(errors.InvalidParameter, op, "missing vault credential repo function")
+	}
+	if logger == nil {
+		return nil, errors.New(errors.InvalidParameter, op, "missing logger")
+	}
+	opts := getOpts(opt...)
+	if opts.withLimit == 0 {
+		// zero signals the boundary defaults should be used.
+		opts.withLimit = db.DefaultLimit
+	}
+	return &TokenRenewalJob{
+		repoFn: repoFn,
+		logger: logger,
+		limit:  opts.withLimit,
+	}, nil
+}
+
+// Status returns the current status of the token renewal job.  Total is the total number
+// of tokens that are set to be returned and completed is the number of tokens already renewed.
+func (r *TokenRenewalJob) Status() scheduler.JobStatus {
+	return scheduler.JobStatus{
+		Completed: r.numProcessed,
+		Total:     r.numTokens,
+	}
+}
+
+// Run queries the vault credential repo for tokens that need to be renewed, it then creates
+// a vault client and renews each token.
+func (r *TokenRenewalJob) Run(ctx context.Context) error {
+	const op = "vault.(TokenRenewalJob).Run"
+
+	// Verify context is not done before running
+	if err := ctx.Err(); err != nil {
+		return errors.Wrap(err, op)
+	}
+
+	repo, err := r.repoFn()
+	if err != nil {
+		return errors.Wrap(err, op)
+	}
+
+	var rps []*renewablePrivateStore
+	err = repo.reader.SearchWhere(ctx, &rps, `now() + (?||'min')::interval > renewal_time`, []interface{}{tokenRenewalWindow}, db.WithLimit(r.limit))
+	if err != nil {
+		return errors.Wrap(err, op)
+	}
+
+	// Set numTokens for status report
+	r.numTokens = len(rps)
+
+	for _, s := range rps {
+		// Verify context is not done before renewing next token
+		if err := ctx.Err(); err != nil {
+			return errors.Wrap(err, op)
+		}
+
+		databaseWrapper, err := repo.kms.GetWrapper(ctx, s.Store.ScopeId, kms.KeyPurposeDatabase)
+		if err != nil {
+			return errors.Wrap(err, op, errors.WithMsg("unable to get database wrapper"))
+		}
+
+		err = s.Store.decrypt(ctx, databaseWrapper)
+		if err != nil {
+			return errors.Wrap(err, op)
+		}
+
+		vc, err := s.Store.client()
+		if err != nil {
+			return errors.Wrap(err, op)
+		}
+
+		var respErr *vault.ResponseError
+		renewedToken, err := vc.renewToken()
+		if ok := errors.As(err, &respErr); ok && respErr.StatusCode == http.StatusForbidden {
+			// Vault returned a 403 when attempting a renew self, the token is either expired
+			// or malformed.  Set status to "expired" so credentials created with token can be
+			// cleaned up.
+			numRows, err := repo.writer.Exec(ctx, updateTokenStatusQuery, []interface{}{StatusExpired, s.Store.TokenHmac})
+			if err != nil {
+				return errors.Wrap(err, op)
+			}
+			if numRows != 1 {
+				return errors.New(errors.Unknown, op, "token expired but failed to update repo")
+			}
+			r.logger.Info("Vault credential store token has expired", "credential store id", s.Store.StoreId, "previous status", s.Store.TokenStatus)
```

Comment:
Perhaps, I might consider only logging if the token that expires is `current`.  I added this log because if a `current` token becomes invalid for whatever reason the end user will need to update the credential store with a new token.